homeassistant/components/mqtt/config_flow.py

```diff
@@ -721,6 +853,294 @@ def _validate(
+
+        data_schema_fields = PLATFORM_MQTT_FIELDS[platform] | COMMON_MQTT_FIELDS
+        data_schema = data_schema_from_fields(data_schema_fields)
+        if user_input is not None:
+            # Test entity fields against the validator
+            validate_user_input(user_input, data_schema_fields, errors)
+            if not errors:
+                self.update_component_fields(data_schema, user_input)
+                self._component_id = None
+                if self.source == SOURCE_RECONFIGURE:
+                    self._dirty = True
+                    return await self.async_step_summary_menu()
+                return self._async_create_entry()
+
+            data_schema = self.add_suggested_values_to_schema(data_schema, user_input)
+        else:
+            data_schema = self.add_suggested_values_to_schema(
+                data_schema, self._subentry_data["components"][self._component_id]
+            )
+
+        return self.async_show_form(
+            step_id="mqtt_platform_config",
+            data_schema=data_schema,
+            description_placeholders={
+                "mqtt_device": device_name,
+                CONF_PLATFORM: platform,
+                "entity": full_entity_name,
+            },
+            errors=errors,
+            last_step=False,
+        )
+
+    @callback
+    def _async_create_entry(
+        self, user_input: dict[str, Any] | None = None
+    ) -> SubentryFlowResult:
+        """Confirm creating a new MQTT device."""
+        device_name = self._subentry_data[CONF_DEVICE][CONF_NAME]
+        component: dict[str, Any] = next(
+            iter(self._subentry_data["components"].values())
+        )
+        platform = component[CONF_PLATFORM]
+        entity_name: str | None
+        if entity_name := component.get(CONF_NAME):
+            full_entity_name: str = f"{device_name} {entity_name}"
+        else:
+            full_entity_name = device_name
+
+        return self.async_create_entry(
+            data=self._subentry_data,
+            title=self._subentry_data[CONF_DEVICE][CONF_NAME],
+            description_placeholders={
+                "entity": full_entity_name,
+                CONF_PLATFORM: platform,
+            },
+        )
+
+    async def async_step_summary_menu(
+        self, user_input: dict[str, Any] | None = None
+    ) -> SubentryFlowResult:
+        """Show summary menu and decide to add more entities or to finish the flow."""
+        self._component_id = None
+        mqtt_device = self._subentry_data[CONF_DEVICE][CONF_NAME]
+        mqtt_items = ", ".join(
+            f"{mqtt_device} {component.get(CONF_NAME, '-')}"
+            for component in self._subentry_data["components"].values()
+        )
+        menu_options = [
+            "entity",
+            "update_entity",
+        ]
+        if len(self._subentry_data["components"]) > 1:
+            menu_options.append("delete_entity")
+        menu_options.append("device")
+        if self._dirty:
```

Comment:
This seems error prone to keep track of and make sure it's updated in the correct places, for example I think it's forgotten in the entity step. Also, the `_dirty` flag is set also if nothing is changed so it's not very accurate.
Instead, I would suggest to make a deep copy of the config entry settings when starting reconfigure, and if there's a difference, show the "Save changes" menu option. Maybe the deep copy is not needed, and we can just compare with `subentry.data`?

Reply:
A compare was added